drivers/pcie/pcie.c

```diff
@@ -46,37 +46,26 @@ void pcie_set_cmd(pcie_bdf_t bdf, uint32_t bits, bool on)
 	pcie_conf_write(bdf, PCIE_CONF_CMDSTAT, cmdstat);
 }
 
-static uint32_t pcie_get_bar(pcie_bdf_t bdf, unsigned int index, bool io)
+uintptr_t pcie_get_mbar(pcie_bdf_t bdf, unsigned int index)
 {
-	int bar;
-	uint32_t data;
-
-	for (bar = PCIE_CONF_BAR0; bar <= PCIE_CONF_BAR5; ++bar) {
-		data = pcie_conf_read(bdf, bar);
-		if (data == PCIE_CONF_BAR_NONE) {
-			continue;
-		}
+	uint32_t reg, bar;
+	uintptr_t addr = PCIE_CONF_BAR_NONE;
 
-		if ((PCIE_CONF_BAR_IO(data) && io) ||
-		    (PCIE_CONF_BAR_MEM(data) && !io)) {
-			if (index == 0) {
-				return PCIE_CONF_BAR_ADDR(data);
-			}
-
-			--index;
+	reg = PCIE_CONF_BAR0;
+	for (bar = 0; bar < index && reg <= PCIE_CONF_BAR5; bar++) {
```

Comment:
One thing to note: this preserves the previous convention where the "index" of the BAR is the natural one, where you have BAR 0, 1, 2, etc...  That requires the use of that loop, because the register entries are variable sized (64 bit entries take two slots).

Note that this is NOT the convention used in Linux (I don't think, though I'd have to recheck to make sure).  There, the bar index just specifies what register slot the BAR starts in, so if you have a device with three 64-bit BAR entries, they would be named "0, 2 and 4".   The advantage of that scheme is that it doesn't require this clumsy loop.

I didn't change it because that might break existing devicetree configurations and I didn't want to hunt them all down to update them.  But if we do want to make this API change it's better done sooner than later.